File: backend/shopping/src/model/order.py
import time
from typing import List

from attr import dataclass

from src.model.customer import Customer
from src.model.payment import Payment
from src.model.product import Product


@dataclass(frozen=True)
class OrderItem:
    product: Product
    quantity: float

    def total(self) -> float:
        return self.product.price * self.quantity


class Order:
    customer: Customer = None
    items: List[OrderItem] = []
    payment: Payment = None
    closed_at: time = None

    def __init__(self, customer: Customer = None):
        self.items = []
        self.customer = customer

    def add_product(self, product: Product, quantity: int):
        self.items.append(OrderItem(product=product, quantity=quantity))

    def total_amount(self) -> float:
        total = 0

        for item in self.items:
            total += item.total()

        return total

    def close(self, closed_at=time.time()):
        self.closed_at = closed_at

File: backend/src/main/kotlin/service/order/Order.kt
package challenge

import challenge.model.PhysicalShipping
import challenge.model.SubscriptionShipping
import java.util.*

class Order(val customer: Customer, val address: Address) {

    val items = mutableListOf<OrderItem>()

    var closedAt: Date? = null
        private set
    var payment: Payment? = null
        private set
    val totalAmount
        get() = items.sumByDouble { it.total }

    fun addProduct(product: Product, quantity: Int) {
        var productAlreadyAdded = items.any { it.product == product }
        if (productAlreadyAdded)
            throw Exception("The product have already been added. Change the amount if you want more.")

        items.add(OrderItem(product, quantity))
    }

    fun pay(method: PaymentMethod) {
        if (payment != null)
            throw Exception("The order has already been paid!")

        if (items.count() == 0)
            throw Exception("Empty order can not be paid!")

        payment = Payment(this, method)

        close()
    }

    fun ship(){
        items.forEach { (item) ->
            run {
                when (item.type) {
                    ProductType.BOOK -> BookShipping().ship(item)
                    ProductType.DIGITAL -> DigitalShipping(customer).ship(item)
                    ProductType.PHYSICAL -> PhysicalShipping().ship(item)
                    ProductType.MEMBERSHIP -> SubscriptionShipping(customer).ship(item)
                }
            }
        }
    }

    private fun close() {
        closedAt = Date()
    }
}